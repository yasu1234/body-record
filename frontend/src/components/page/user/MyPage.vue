<script setup>
import { ref, onMounted } from "vue";
import { useRouter, useRoute } from "vue-router";
import { useToast } from "primevue/usetoast";
import { toastService } from "../../../js/toast.js";
import Toast from "primevue/toast";
import { axiosInstance, setupInterceptors } from "../../../js/axios.js";

import Header from "../../layout/Header.vue";
import RecordCard from "../../layout/RecordCard.vue";
import KnowledgeCard from "../../layout/KnowledgeCard.vue";
import Chart from "../../atom/Chart.vue";
import MonthPicker from "../../atom/MonthPicker.vue";
import Profile from "../../layout/Profile.vue";
import TabMenu from "../../layout/TabMenu.vue";

const route = useRoute();
const router = useRouter();
const toast = useToast();
const toastNotifications = new toastService(toast);
setupInterceptors(router);

const user = ref(null);
const support = ref(null);
const records = ref([]);
const knowledges = ref([]);
const isMoreRecords = ref(false);
const isMoreKnowledges = ref(false);
const userId = ref(0);
const month = ref({
  month: new Date().getMonth(),
  year: new Date().getFullYear(),
});

onMounted(() => {
  checkLogin();
});

var weigtData = ref({
  labels: [],
  datasets: [
    {
      label: "体重(kg)",
      data: [],
      borderColor: "rgb(75, 192, 192)",
    },
  ],
});

var fatPercentageData = ref({
  labels: [],
  datasets: [
    {
      label: "体脂肪率(%)",
      data: [],
      borderColor: "rgb(149, 101, 255)",
    },
  ],
});

const checkLogin = async () => {
  try {
    const res = await axiosInstance.get("/api/v1/users/check_login");
    userId.value = res.data.user.id;
    getProfile();
    getSupportCount();
    getUserRecord();
    getUserKnowledge();
    getMonthRecord();
  } catch (error) {
    toastNotifications.displayError("ユーザー情報を取得できません", "");
  }
};

const getProfile = async () => {
  try {
    const res = await axiosInstance.get(`/api/v1/profiles/${userId.value}`);

    user.value = res.data.user;
  } catch (error) {
    user.value = null;
    toastNotifications.displayError("プロフィールの取得に失敗しました", "");
  }
};

const getSupportCount = async () => {
  try {
    const res = await axiosInstance.get(`/api/v1/support_counts`, {
      params: {
        user_id: userId.value,
      },
    });

    support.value = res.data.user;
  } catch (error) {
    support.value = null;
  }
};

const getUserRecord = async () => {
  try {
    const res = await axiosInstance.get(`/api/v1/user_records`, {
      params: {
        user_id: userId.value,
      },
    });

    records.value = res.data.records;
    isMoreRecords.value = res.data.is_more;
  } catch (error) {
    records.value = [];
    isMoreRecords.value = false;
  }
};

const getMonthRecord = async () => {
  try {
    const res = await axiosInstance.get(`/api/v1/graph_records`, {
      params: {
        user_id: userId.value,
        targetYear: month.value.year,
        targetMonth: month.value.month + 1,
      },
    });

    weigtData.value.labels = res.data.records.map(
      (record) => record.graph_formatted_date
    );
    weigtData.value.datasets[0].data = res.data.records.map(
      (record) => record.weight
    );
    fatPercentageData.value.labels = res.data.records.map(
      (record) => record.graph_formatted_date
    );
    fatPercentageData.value.datasets[0].data = res.data.records.map(
      (record) => record.fat_percentage
    );
  } catch (error) {}
};

const getUserKnowledge = async () => {
  try {
    const res = await axiosInstance.get(`/api/v1/user_knowledges`, {
      params: {
        user_id: userId.value,
      },
    });

    knowledges.value = res.data.knowledges;
    isMoreKnowledges.value = res.data.is_more;
  } catch (error) {
    knowledges.value = [];
    isMoreKnowledges.value = false;
  }
};

function monthChange(event) {
  month.value = event;
  getMonthRecord();
}

const clickRecord = (item) => {
  router.push({ name: "RecordDetail", params: { id: item.id } });
};

const showMoreRecords = () => {
  router.push({ name: "Home" });
};
</script>

<template>
  <Toast position="top-center" />
  <Header />
  <TabMenu />
  <Profile :user="user" :support="support" />
  <div class="weight-graph">
    <Chart :data="weigtData" />
  </div>
  <div class="weight-graph">
    <Chart :data="fatPercentageData" />
  </div>
  <MonthPicker
    :date="month"
    @update:month="monthChange"
    class="graph-month-picker"
  />
  <div class="text-center">
    <div class="pt-10">
      <span class="section-title">投稿した記録</span>
    </div>
    <div class="mt-5">
      <RecordCard
        v-if="records.length > 0"
        v-for="record in records"
        v-bind="record"
        :record="record"
        @recordClick="clickRecord(record)"
      />
      <p v-else>登録された記録はありません</p>
    </div>
    <button
      v-if="isMoreRecords"
      @click="showMoreRecords"
      class="more-show-button mt-5"
    >
      もっと見る
    </button>
    <div class="pt-10">
      <span class="section-title">投稿した知識</span>
    </div>
    <div class="mt-5">
      <KnowledgeCard
        v-if="knowledges.length > 0"
        v-for="knowledge in knowledges"
        v-bind="knowledge"
        :knowledgeTitle="knowledge.title"
        :knowledgeContent="knowledge.content"
      />
      <p v-else>記事を作成していません</p>
    </div>
    <button
      v-if="isMoreKnowledges"
      @click="showMoreRecords"
      class="more-show-button mt-5"
    >
      もっと見る
    </button>
  </div>
</template>

<style scoped>
.section-title {
  border-bottom: solid 5px #ffa500;
  font-size: 25px;
  font-weight: bold;
}
.weight-graph {
  margin-top: 30px;
  width: 600px;
  margin-left: auto;
  margin-right: auto;
}
.graph-month-picker {
  margin-top: 20px;
  width: 600px;
  margin-left: auto;
  margin-right: auto;
}
.more-show-button {
  background: #fff;
  color: #000000;
  border: 1px solid #ccc;
  font-size: 16px;
  font-weight: bold;
  cursor: pointer;
  padding: 10px 50px;
}

@media screen and (max-width: 768px) {
  .weight-graph {
    width: auto;
    margin-left: 20px;
    margin-right: 20px;
    margin-top: 30px;
  }
  .graph-month-picker {
    margin-top: 20px;
    width: auto;
    margin-left: 20px;
    margin-right: 20px;
  }
}
</style>
