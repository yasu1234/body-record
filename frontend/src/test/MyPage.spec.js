import { shallowMount } from "@vue/test-utils";
import MyPage from "@/components/page/user/MyPage.vue";

jest.mock("axios");
jest.mock("vue-router", () => {
  const router = {
    push: jest.fn(),
    pathname: "/",
    route: "",
    asPath: "",

    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn(),
    },
    onBeforeRouteUpdate: jest.fn(),
  };
  return {
    useRouter: jest.fn().mockReturnValue(router),
    useRoute: jest.fn().mockReturnValue({
      params: {
      },
      query: {
      },
    }),
    onBeforeRouteUpdate: jest.fn(),
  };
});

jest.mock("primevue/usetoast", () => ({
  useToast: jest.fn(() => ({
    add: jest.fn(),
    removeAll: jest.fn(),
  })),
}));

jest.mock("@/js/axios.js", () => ({
  axiosInstance: {
    get: jest.fn(),
    interceptors: {
      request: { use: jest.fn() },
      response: { use: jest.fn() },
    },
  },
  setupInterceptors: jest.fn(),
}));

describe("MyPage", () => {
  let wrapper;
  let mockedAxiosInstance;

  beforeEach(() => {
    jest.mock("@/js/axios.js", () => ({
      axiosInstance: {
        get: jest.fn(),
        interceptors: {
          request: { use: jest.fn() },
          response: { use: jest.fn() },
        },
      },
      setupInterceptors: jest.fn(),
    }));

    // モックされたaxiosInstanceを取得
    const axiosModule = require("@/js/axios.js");
    mockedAxiosInstance = axiosModule.axiosInstance;

    // onMounted内でaxios.getが呼ばれるため、このタイミングでモック設定
    mockedAxiosInstance.get.mockResolvedValue({
      data: {},
    });

    wrapper = shallowMount(MyPage, {
      global: {
        stubs: ["router-link", "router-view"],
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("SnapShotテスト初期表示(データなし)", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});

describe("MyPage", () => {
  let wrapper;
  let mockedAxiosInstance;

  beforeEach(() => {
    jest.mock("@/js/axios.js", () => ({
      axiosInstance: {
        get: jest.fn(),
        interceptors: {
          request: { use: jest.fn() },
          response: { use: jest.fn() },
        },
      },
      setupInterceptors: jest.fn(),
    }));

    // モックされたaxiosInstanceを取得
    const axiosModule = require("@/js/axios.js");
    mockedAxiosInstance = axiosModule.axiosInstance;

    // onMounted内でaxios.getが呼ばれるため、このタイミングでモック設定
    mockedAxiosInstance.get.mockResolvedValue({
      data: {
        user: {
          id: 1,
          name: "TestUser",
        },
        knowledges: [
          { id: 1, title: "Test1", content: "TESTTEST1" },
          { id: 2, title: "Test2", content: "TESTTEST2" },
        ],
        records: [
            {id: 1, date: "2024-06-24", fat_percentage: 12.1},
            {id:2, name: "2024-06-25", weight: 65}
        ],
        is_more: true
      },
    });

    wrapper = shallowMount(MyPage, {
      global: {
        stubs: ["router-link", "router-view"],
      },
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("SnapShotテスト初期表示(データあり)", () => {
    expect(wrapper.element).toMatchSnapshot();
  });
});
