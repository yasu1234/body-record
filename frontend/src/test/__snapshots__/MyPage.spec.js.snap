// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyPage SnapShotテスト初期表示(データあり) 1`] = `
<div
  data-v-app=""
>
  
  <toast-stub
    position="top-center"
  />
  <header-stub />
  <tab-menu-stub />
  <profile-stub
    support="[object Object]"
    user="[object Object]"
  />
  <div
    class="weight-graph"
  >
    <chart-stub
      data="[object Object]"
    />
  </div>
  <div
    class="weight-graph"
  >
    <chart-stub
      data="[object Object]"
    />
  </div>
  <month-picker-stub
    class="graph-month-picker"
    date="[object Object]"
  />
  <div
    class="text-center"
  >
    <div
      class="pb-5"
    >
      <div
        class="pt-10"
      >
        <span
          class="section-title"
        >
          作成した記録
        </span>
      </div>
      <div
        class="mt-5"
      >
        
        <div>
          <record-card-stub
            record="[object Object]"
          />
        </div>
        <div>
          <record-card-stub
            record="[object Object]"
          />
        </div>
        
      </div>
      <button
        class="more-show-button mt-5"
      >
         もっと見る 
      </button>
    </div>
    <div
      class="pb-8"
    >
      <div
        class="pt-10"
      >
        <span
          class="section-title"
        >
          投稿した記事
        </span>
      </div>
      <div
        class="mt-5"
      >
        
        <div>
          <knowledge-card-stub
            knowledge="[object Object]"
          />
        </div>
        <div>
          <knowledge-card-stub
            knowledge="[object Object]"
          />
        </div>
        
      </div>
      <button
        class="more-show-button mt-5"
      >
         もっと見る 
      </button>
    </div>
  </div>
  
</div>
`;

exports[`MyPage SnapShotテスト初期表示(データなし) 1`] = `
<div
  data-v-app=""
>
  
  <toast-stub
    position="top-center"
  />
  <header-stub />
  <tab-menu-stub />
  <profile-stub />
  <div
    class="weight-graph"
  >
    <chart-stub
      data="[object Object]"
    />
  </div>
  <div
    class="weight-graph"
  >
    <chart-stub
      data="[object Object]"
    />
  </div>
  <month-picker-stub
    class="graph-month-picker"
    date="[object Object]"
  />
  <div
    class="text-center"
  >
    <div
      class="pb-5"
    >
      <div
        class="pt-10"
      >
        <span
          class="section-title"
        >
          作成した記録
        </span>
      </div>
      <div
        class="mt-5"
      >
        <p>
          作成された記録はありません
        </p>
      </div>
      <!--v-if-->
    </div>
    <div
      class="pb-8"
    >
      <div
        class="pt-10"
      >
        <span
          class="section-title"
        >
          投稿した記事
        </span>
      </div>
      <div
        class="mt-5"
      >
        <p>
          記事を投稿していません
        </p>
      </div>
      <!--v-if-->
    </div>
  </div>
  
</div>
`;
